test/e2e/autotls/auto_tls_test.go

```diff
@@ -43,18 +43,64 @@ import (
 	v1test "knative.dev/serving/test/v1"
 )
 
-// To run this test locally with cert-manager, you need to
-// 1. Install cert-manager from `third_party/` directory.
-// 2. Run the command below to do the configuration:
-//    kubectl apply -f test/config/autotls/certmanager/selfsigned/
-func TestPerKsvcCertLocalCA(t *testing.T) {
+const dnsRecordDeadlineSec = 600 // 10 min
+
+type dnsRecord struct {
+	ip     string
+	domain string
+}
+
+type config struct {
+	// ServiceName is the name of testing Knative Service.
+	// It is not required for self-signed CA or for the HTTP01 challenge when wildcard domain
+	// is mapped to the Ingress IP.
+	TLSServiceName string `envconfig:"tls_service_name" required: "false"`
+	// AutoTLSTestName is the name of the auto tls. It is not required for local test.
+	AutoTLSTestName string `envconfig:"auto_tls_test_name" required: "false"`
+}
+
+var env config
+
+// To run this E2E test locally:
+// test case 1: testing per ksvc certificate provision with self-signed CA
+//   1) `kubectl label namespace serving-tests networking.internal.knative.dev/disableWildcardCert=true`
+//   2) `kubectl delete kcert --all -n serving-tests`
+//   3) `kubectl apply -f test/config/autotls/certmanager/selfsigned/`
+//   4) `go test -v -tags=e2e -count=1 -timeout=600s ./test/e2e/autotls/... -run  ^TestAutoTLS$`
+// test case 2: testing per namespace certificate provision with self-signed CA
+//   1) `kubectl delete kcert --all -n serving-tests`
+//   2) kubectl apply -f test/config/autotls/certmanager/selfsigned/
+//   3) Run `kubectl edit namespace serving-tests` and remove the label networking.internal.knative.dev/disableWildcardCert
+//   4) `go test -v -tags=e2e -count=1 -timeout=600s ./test/e2e/autotls/... -run  ^TestAutoTLS$`
+// test case 3: testing per ksvc certificate provision with HTTP challenge
+//   1) `kubectl label namespace serving-tests networking.internal.knative.dev/disableWildcardCert=true`
+//   2) `kubectl delete kcert --all -n serving-tests`
+//   3) `kubectl apply -f test/config/autotls/certmanager/http01/`
+//   4) `export SERVICE_NAME=http01`
+//   5) `kubectl patch cm config-domain -n knative-serving -p '{"data":{"<your-custom-domain>":""}}'`
+//   6) Add a DNS A record to map host `http01.serving-tests.<your-custom-domain>` to the Ingress IP.
+//   7) `go test -v -tags=e2e -count=1 -timeout=600s ./test/e2e/autotls/... -run  ^TestAutoTLS$`
+func TestAutoTLS(t *testing.T) {
+	if err := envconfig.Process("", &env); err != nil {
+		t.Fatalf("Failed to process environment variable: %v.", err)
+	}
+	testName := "TestAutoTLS"
+	if len(env.AutoTLSTestName) != 0 {
+		testName = env.AutoTLSTestName
```

Comment:
This logic doesn't make sense, could you comment it?

Reply:
@n3wscott can envconfig do defaults?

Reply:
Looks like it can: https://github.com/kelseyhightower/envconfig#struct-tag-support